Imports from the same package should be locked to the exact version number

jasongrout opened the pull request:
There is no guarantee that internal package imports will be compatible across patch versions, so we must lock to the exact version number for internal imports.

jasongrout committed:
Imports from the same package should be locked to the exact version number.

There is no guarantee that internal package imports will be compatible across patch versions, so we must lock to the exact version number for internal imports.

jasongrout commented:
@blink1073 - I think this is ready for review.

jasongrout commented:
Actually, it seems that something else needs to be fixed. For some reason, *some* local imports are affected by this, but others aren't. I think it may be a cache invalidation issue. Investigating...

jasongrout commented:
I think what may be happening is that the semver mangled paths may be cached at the chunk level rather than the module level. Maybe you can show me again how to have nice debugging in webpack at the JupyterLab meeting...

blink1073 commented:
```bash
npm install -g devtool`.
devtool <my_build_script.js>
```

blink1073 commented:
or, if running webpack directly,

`devtool <path_to_webpack_in_node_modules/bin>`

jasongrout commented:
Cool, thanks @blink1073!

jasongrout commented:
Wow, it works! Thanks again, @blink1073.

jasongrout committed:
Change the semver range resolution to be explicit about the module requiring the dependency.

Apparently webpack’s dependency module metadata contains an issuer attribute, but the issuer attribute only seems to indicate *one* of the modules requiring the dependency, which may not be the issuer under consideration. Because our semver range resolution algorithm depends on the issuer, we must pass the issuer in explicitly.

jasongrout committed:
Local package versions should resolve the version of the dependency, not the issuer.

jasongrout commented:
@blink1073 - I think this is ready for review. You'll notice if you generate the jupyterlab bundle, for example, before this change there are many instances where files from the same library have different semver ranges (for example, side-by-side imports of phosphor modules will have some modules with semver range `~` and others with semver range `^`).

I think this PR takes care of this problem, as documented now in the readme.

blink1073 commented:
Looks good, tested against JLab master, and agreed that the `require()` statements are much more uniform.

blink1073 commented:
Thanks!

jasongrout commented:
Thanks!
